Fixes kubernetes/kubernetes#29271: accept prefixed namespaces

erhudy opened the pull request:
**What this PR does / why we need it**: `kubectl get namespaces -o name` outputs the names of all namespaces, prefixed with `namespaces/`. This changeset allows these namespace names to be passed directly back in to `kubectl` via the `-n` flag without reprocessing them to remove `namespaces/`.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #29271

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```NONE
```

liggitt commented on pkg/kubectl/cmd/get.go:
```diff
@@ -178,6 +178,10 @@ func RunGet(f cmdutil.Factory, out, errOut io.Writer, cmd *cobra.Command, args [
 	}
 
+	if strings.HasPrefix(cmdNamespace, "namespaces/") {
+		cmdNamespace = strings.Split(cmdNamespace, "/")[1]
```
TrimPrefix would be preferable, since splitting on `/` could accidentally allow accepting `namespaces/foo/bar` as `foo`

liggitt replied:
also, this should be done in `DefaultNamespace`, not in one specific command. the behavior of `-n ...` should be consistent

xiangpengzhao replied:
Also, I'm afraid `cmdNamespace = strings.Split(cmdNamespace, "/")[1]` will cause a panic if `cmdNamespace` is specified to `namespaces/`.

xiangpengzhao replied:
@erhudy could you please add a test case to prove that?

liggitt commented:
cc @kubernetes/sig-cli-pr-reviews 

if `-n` accepted resource-qualified syntax, I'd sort of expect all of them to work... `namespace/foo`, `namespaces/foo`, `ns/foo`, etc.

xiangpengzhao commented on pkg/kubectl/cmd/get_test.go:
```diff
@@ -269,6 +269,32 @@ func TestGetObjects(t *testing.T) {
+		NegotiatedSerializer: unstructuredSerializer,
+		Resp:                 &http.Response{StatusCode: 200, Header: defaultHeader(), Body: objBody(codec, &pods.Items[0])},
+	}
+	tf.Namespace = "namespaces/test"
```
Hmmm, I change this line to `tf.Namespace = "namespaces/"` and get a panic. See my comment above :)

```
--- FAIL: TestGetObjectsWithPrefixedNamespace (0.00s)
panic: Fail in goroutine after TestEdit has completed [recovered]
	panic: Fail in goroutine after TestEdit has completed

goroutine 452 [running]:
testing.tRunner.func1(0xc420385040)
	/usr/local/go/src/testing/testing.go:622 +0x29d
panic(0x19d2000, 0xc42031fca0)
	/usr/local/go/src/runtime/panic.go:489 +0x2cf
testing.(*common).Fail(0xc420385ad0)
	/usr/local/go/src/testing/testing.go:445 +0x116
testing.(*common).Errorf(0xc420385ad0, 0x1ccd96c, 0x25, 0xc4208a7b88, 0x4, 0x4)
	/usr/local/go/src/testing/testing.go:518 +0x83
k8s.io/kubernetes/pkg/kubectl/cmd.TestEdit.func5(0xc420af9e50, 0x44, 0x1)
	/home/cloud/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/edit_test.go:264 +0x203
k8s.io/kubernetes/pkg/kubectl/cmd/util.checkErr(0x2a90000, 0xc42031fc10, 0xc42092cf00)
	/home/cloud/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/util/helpers.go:178 +0x948
k8s.io/kubernetes/pkg/kubectl/cmd/util.CheckErr(0x2a90000, 0xc42031fc10)
	/home/cloud/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/util/helpers.go:119 +0x41
k8s.io/kubernetes/pkg/kubectl/cmd.NewCmdGet.func1(0xc420270480, 0xc42023b6e0, 0x2, 0x2)
	/home/cloud/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/get.go:122 +0xad
k8s.io/kubernetes/pkg/kubectl/cmd.TestGetObjectsWithPrefixedNamespace(0xc420385040)
	/home/cloud/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/get_test.go:288 +0x455
testing.tRunner(0xc420385040, 0x1d4fdc0)
	/usr/local/go/src/testing/testing.go:657 +0x96
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:697 +0x2ca
FAIL	k8s.io/kubernetes/pkg/kubectl/cmd	7.986s
make: *** [test] Error 1
```

xiangpengzhao reviewed (changes requested):
please also fix the potential panic.

erhudy commented:
Revised per feedback. The check has been revised to use `strings.TrimPrefix` instead of `strings.Split`, has been moved to the factory implementation, and two tests have been added there to validate the namespace.

Per the concern about the panic, I found that after the change to `strings.TrimPrefix`, specifying `-n namespaces/` would result in this somewhat cryptic error message:
```
	the namespace from the provided object "kube-system" does not match the namespace "". You must pass '--namespace=kube-system' to perform this operation.
```

Specifying `-n ''` did not have the same effect. I added an extra check in the factory implementation to return the simple error **namespace cannot be empty** if somebody passes `-n namespaces/`.

liggitt commented on pkg/kubectl/cmd/util/factory_client_access.go:
```diff
@@ -451,7 +451,16 @@ func (f *ring0Factory) Resumer(info *resource.Info) ([]byte, error) {
 
 func (f *ring0Factory) DefaultNamespace() (string, bool, error) {
-	return f.clientConfig.Namespace()
+	namespace, boolRetval, err := f.clientConfig.Namespace()
```
hrmm... here be dragons... leaving an invalid value being returned from clientConfig.Namespace() is no good. for example, see this logic that checks whether the namespace was passed and should override:
```
func (config *DirectClientConfig) Namespace() (string, bool, error) {
	if config.overrides != nil && config.overrides.Context.Namespace != "" {
		// In the event we have an empty config but we do have a namespace override, we should return
		// the namespace override instead of having config.ConfirmUsable() return an error. This allows
		// things like in-cluster clients to execute `kubectl get pods --namespace=foo` and have the
		// --namespace flag honored instead of being ignored.
		return config.overrides.Context.Namespace, true, nil
	}
```

this has to go even deeper, probably into the Namespace() method, and/or possibly into the getContext() method. It's also an open question whether we should accept the `namespace/...` syntax everywhere (e.g. in kubeconfig files, etc) or just via the command-line `-n` flag

erhudy replied:
My intent was just to allow it to be valid in `-n`. Would it be an acceptable compromise to throw a fatal error inside here if the result of the trim operation ends up being empty, since that's the only error condition?

liggitt replied:
no, it really has to be handled deeper... in the case where they specify a "valid" value of `namespaces/foo`, `clientConfig.Namespace()` would return a value that would be invalid to use directly as a namespace.

erhudy replied:
Okay, I will look at pushing it up to the client configuration.

shiywang replied:
yes, and having a generic namespace trim function/logic would be good.

erhudy commented:
Revised again to move the changes into client-go (intentionally only applying to overrides, i.e. supplied through `-n`, and not to the full configuration context).

erhudy commented:
Pinging @liggitt and @xiangpengzhao for reconsideration.

dims commented:
/ok-to-test

erhudy commented:
Pinging @liggitt @xiangpengzhao @smarterclayton @dchen1107 on this PR.

xiangpengzhao commented on staging/src/k8s.io/client-go/tools/clientcmd/client_config.go:
```diff
@@ -296,12 +296,21 @@ func canIdentifyUser(config restclient.Config) bool {
-	if config.overrides != nil && config.overrides.Context.Namespace != "" {
+	// if an override is provided and it starts with namespaces/ (as generated by `kubectl get ns -o name`), strip it off
+	// (this is intentionally targeted towards interactive use only)
+	namespace := config.overrides.Context.Namespace
```
should check `if config.overrides != nil` before this line, or you may get a panic.

xiangpengzhao commented on staging/src/k8s.io/client-go/tools/clientcmd/client_config.go:
```diff
@@ -296,12 +296,21 @@ func canIdentifyUser(config restclient.Config) bool {
+		namespace = strings.TrimPrefix(namespace, namespacesPrefix)
+	}
+
+	if config.overrides != nil && namespace != "" {
```
then the condition `config.overrides != nil` can be removed here.

xiangpengzhao commented on staging/src/k8s.io/client-go/tools/clientcmd/client_config_test.go:
```diff
@@ -526,3 +527,43 @@ func TestNamespaceOverride(t *testing.T) {
 	matchStringArg("foo", ns, t)
 }
+
+func TestNamespacePrefixStrip(t *testing.T) {
```
Maybe the two test funcs can be in one single test func with multiple test cases. Anyway, this is ok as-is.

erhudy commented:
@xiangpengzhao Revised to improve the nil check on `config.overrides`.

erhudy commented:
/test pull-kubernetes-e2e-gce-bazel

liggitt committed:
Allow -n namespaces/<ns>

liggitt commented on staging/src/k8s.io/client-go/tools/clientcmd/client_config.go:
```diff
@@ -296,12 +296,23 @@ func canIdentifyUser(config restclient.Config) bool {
 
 // Namespace implements ClientConfig
 func (config *DirectClientConfig) Namespace() (string, bool, error) {
-	if config.overrides != nil && config.overrides.Context.Namespace != "" {
```
I think this should go even deeper, down into the flag binding itself. We really only want this special handling to apply to the namespace flag itself (not necessarily the override struct), and there are other configs (like DeferredLoadingClientConfig that make use of overrides that would miss any special logic here).

Essentially, we want the data in the overrides struct to be good, not need fixing up. That could be done with a transforming string flag that lets us strip the namespace prefix prior to the value getting stored in the overrides struct - see https://github.com/liggitt/kubernetes/commit/e2478d3fb7c51d2615d10a1134a967a25eed7365 as an example

erhudy committed:
Add tests for stripping "namespaces/" from passed-in namespace

erhudy commented:
@liggitt Okay. I cherry-picked your commit onto my branch (to maintain your authorship) and then rewrote the tests so that they're testing values passed in via the `--namespace` flag.

liggitt commented on staging/src/k8s.io/client-go/tools/clientcmd/overrides_test.go:
```diff
@@ -0,0 +1,55 @@
+	}
+
+	for before, after := range testData {
+		BindOverrideFlags(&ConfigOverrides{}, pflag.CommandLine, RecommendedConfigOverrideFlags(""))
```
no need to use `pflag.CommandLine`, can just use a local `fs := &pflag.FlagSet{}`

liggitt replied:
I'd recommend a local `overrides := &ConfigOverrides{}` as well, and check `overrides.Context.Namespace` rather than `pflag.CommandLine.GetString("namespace")`

erhudy committed:
Update Bazel configuration for flag.go and overrides_test.go

liggitt commented:
/retest

erhudy commented:
/test pull-kubernetes-federation-e2e-gce

liggitt commented:
/retest

erhudy commented:
Waiting for https://github.com/kubernetes/kubernetes/issues/51692 to be fixed and then resubmitting.

liggitt commented:
no further comments from me, would like LGTM from @kubernetes/sig-cli-pr-reviews since one of the commits is now made by me

ericchiang commented:
Logic looks good and it looks like the earlier potential panics were fixed.

/lgtm

liggitt commented:
/approve
